src/global/reco/MatchClusters_factory.cc

```diff
@@ -23,36 +23,27 @@ namespace eicrecon {
         InitLogger(GetApplication(), GetPrefix(), "info");
 
         m_match_algo.init(m_log);
-
-        // Set up association tags
-        for (const std::string& input_tag : GetInputTags()) {
-            // "EcalEndcapNClusters" => "EcalEndcapNClusterAssociations"
-            auto pos = input_tag.find("Clusters");
-
-            // Validate that our input collection names conform to this convention
-            if (pos == std::string::npos) {
-                throw std::runtime_error(fmt::format("input_tag=\"{}\" doesn't end with \"Clusters\"", input_tag));
-            }
-            std::string output_tag = input_tag;
-            output_tag.replace(pos, 8, "ClusterAssociations");
-            m_input_assoc_tags.push_back(output_tag);
-        }
     }
 
     void MatchClusters_factory::BeginRun(const std::shared_ptr<const JEvent> &event) {
         // Nothing to do here
     }
 
     void MatchClusters_factory::Process(const std::shared_ptr<const JEvent> &event) {
-        // TODO make input tags changeable
-        auto mc_particles = static_cast<const edm4hep::MCParticleCollection*>(event->GetCollectionBase("MCParticles"));
-        auto charged_particles = static_cast<const edm4eic::ReconstructedParticleCollection*>(event->GetCollectionBase("ReconstructedChargedParticles"));
-        auto charged_particle_assocs = static_cast<const edm4eic::MCRecoParticleAssociationCollection*>(event->GetCollectionBase("ReconstructedChargedParticleAssociations"));
+        size_t tag_ix = 0;
+        auto mc_particles = static_cast<const edm4hep::MCParticleCollection*>(event->GetCollectionBase(GetInputTags()[tag_ix++]));
+        auto charged_particles = static_cast<const edm4eic::ReconstructedParticleCollection*>(event->GetCollectionBase(GetInputTags()[tag_ix++]));
+        auto charged_particle_assocs = static_cast<const edm4eic::MCRecoParticleAssociationCollection*>(event->GetCollectionBase(GetInputTags()[tag_ix++]));
 
         std::vector<const edm4eic::ClusterCollection*> cluster_collections;
         std::vector<const edm4eic::MCRecoClusterParticleAssociationCollection*> cluster_assoc_collections;
 
-        for(auto &input_tag: GetInputTags()) {
+        // FIXME Use proper stride and adjacent views in C++23
+        auto stride = [](int n) {
```

Comment:
I'm actually not familiar with ranges! So for my own understanding, what's going on with this block?

Reply:
Since we only have C++20 support we can't use strided views yet (i.e. every n'th element) which is a C++23 feature. We do have filtered views in C++20, so this creates a filter predicate that is equivalent to a strided view: stride(n) will be true for each n'th call.

Reply:
Ah! I see! So then the block below uses that pick up every other collection starting at tag_ix (i.e. the clusters). And then likewise for the associations. Thanks!!